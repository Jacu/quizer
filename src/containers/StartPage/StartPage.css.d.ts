export const StartPage: string;
export const Menu: string;
export const settingInputs: string;
export const Title: string;
export const quizerIntro: string;
export const SubTitle: string;
