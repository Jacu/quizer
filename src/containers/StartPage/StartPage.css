.StartPage{
    position: absolute;
    background: white;
    width: 60%;    
    height: 50%;
    display: flex;    
    justify-content: center;    
    align-items: center;
    flex-direction: column;
    align-self: center;    
    border-radius: 50px;
    box-shadow: 2px 2px 10px 2px rgba(0,0,0,0.5);
}

.settings {
    width: 100%;
}

.settings ul {
    margin: 50px 0;
    display: flex;
    flex-direction: row;
    list-style: none;
    justify-content: space-around;
}

h1 {
    margin: 50px 0;
}

li {
    flex:1;
}