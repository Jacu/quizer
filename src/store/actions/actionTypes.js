export const INIT = 'INIT';

export const FETCH_CATEGORIES_START = 'FETCH_CATEGORIES_START';
export const FETCH_CATEGORIES_SUCCESS = 'FETCH_CATEGORIES_SUCCESS';
export const FETCH_CATEGORIES_FAIL = 'FETCH_CATEGORIES_FAIL';

export const SET_SETTING = 'SET_SETTING';

export const START_QUIZ = 'START_QUIZ';

export const SHUFFLE_ANSWERS = 'SHUFFLE_ANSWERS';

// ----------------------------

export const FETCH_QUESTIONS_START = 'FETCH_QUESTIONS_START';
export const FETCH_QUESTIONS_SUCCESS = 'FETCH_QUESTIONS_SUCCESS';
export const FETCH_QUESTIONS_FAIL = 'FETCH_QUESTIONS_FAIL';

export const NEXT_QUESTION = 'NEXT_QUESTION';
export const PREV_QUESTION = 'PREV_QUESTION';

export const PICK_ANSWER = 'PICK_ANSWER';