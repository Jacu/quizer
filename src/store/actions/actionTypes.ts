import { Category } from "../reducers/startPage";
import { Dispatch } from "redux";

export const INIT = 'INIT';

export const FETCH_CATEGORIES_START = 'FETCH_CATEGORIES_START';
export const FETCH_CATEGORIES_SUCCESS = 'FETCH_CATEGORIES_SUCCESS';
export const FETCH_CATEGORIES_FAIL = 'FETCH_CATEGORIES_FAIL';

export const SET_SETTING = 'SET_SETTING';

export const GENERATE_URL = 'GENERATE_URL';

export const RESET_START_PAGE = 'RESET_START_PAGE';

// ----------------------------

export const QUIZ_STARTED = 'QUIZ_STARTED';

export const FETCH_QUESTIONS_START = 'FETCH_QUESTIONS_START';
export const FETCH_QUESTIONS_SUCCESS = 'FETCH_QUESTIONS_SUCCESS';
export const FETCH_QUESTIONS_FAIL = 'FETCH_QUESTIONS_FAIL';

export const SHUFFLE_ANSWERS = 'SHUFFLE_ANSWERS';

export const PICK_ANSWER = 'PICK_ANSWER';

export const NEXT_QUESTION = 'NEXT_QUESTION';
export const PREV_QUESTION = 'PREV_QUESTION';

export const QUIZ_ENDED = 'QUIZ_ENDED';
export const QUIZ_QUIT = 'QUIZ_QUIT';

export const CALCULATE_SCORE = 'CALCULATE_SCORE';

export const RESET_QUIZ = "RESET_QUIZ";

// ----------------------------

export interface fetchQuestionsStart {
    type: typeof FETCH_QUESTIONS_START,
}

export interface fetchQuestionsSuccess {
    type: typeof FETCH_QUESTIONS_SUCCESS,
    questions: string,
}

export interface fetchQuestionsFail<T> {
    type: typeof FETCH_QUESTIONS_FAIL,
    error: T,
}

export interface shuffleAnswers {
    type: typeof SHUFFLE_ANSWERS,
}

export interface quizStarted {
    type: typeof QUIZ_STARTED,
}

export interface pickAnswer {
    type: typeof PICK_ANSWER,
    index: number,
    answer: number,
}

export interface nextQuestion {
    type: typeof NEXT_QUESTION,
}

export interface prevQuestion {
    type: typeof PREV_QUESTION,
}

export interface endQuiz {
    type: typeof QUIZ_ENDED,
}

export interface quizQuit {
    type: typeof QUIZ_QUIT,
}

export interface calculateScore {
    type: typeof CALCULATE_SCORE,
}

export interface resetQuiz {
    type: typeof RESET_QUIZ,
}

export type QuizActions = fetchQuestionsStart | fetchQuestionsSuccess | fetchQuestionsFail<any> | shuffleAnswers |
quizStarted | pickAnswer | nextQuestion | prevQuestion | endQuiz | quizQuit | calculateScore | resetQuiz;

// ----------------------------

export interface init extends Dispatch { }

export interface fetchCategoriesStart {
    type: typeof FETCH_CATEGORIES_START,
}

export interface fetchCategoriesSuccess {
    type: typeof FETCH_CATEGORIES_SUCCESS,
    categories: Category[],
}

export interface fetchCategoriesFail<T> {
    type: typeof FETCH_CATEGORIES_FAIL,
    error: T,
}

export interface setSetting {
    type: typeof SET_SETTING,
    setting: string,
    value: string,
}

export interface generateURL {
    type: typeof GENERATE_URL,
}

export interface resetStartPage {
    type: typeof RESET_START_PAGE,
}

export type StartPageActions = fetchCategoriesStart | fetchCategoriesSuccess | fetchCategoriesFail<any> | setSetting | generateURL | resetStartPage;


export type AllActions = QuizActions | StartPageActions;