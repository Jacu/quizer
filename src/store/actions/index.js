export {
    initQuiz,
    pickAnswer,
    nextQuestion,
    prevQuestion,
    quizEnded,
    reset
} from './quiz'
export {
    init,
    setSetting,
    generateURL
} from './startPage'