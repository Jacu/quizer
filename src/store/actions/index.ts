export {
    initQuiz,
    endQuiz,
    reset,
    quizQuit,
} from './quiz'
export {
    init,
    setQuestionAmount,
    setQuestionCategory,
    setQuestionType,
    setQuestionDifficulty,
    generateURL,
} from './startPage'