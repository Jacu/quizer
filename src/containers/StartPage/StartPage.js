import React, { Component } from 'react';
import styles from './StartPage.css';
import { connect } from 'react-redux';
import Input from '../../components/Input/Input';
import * as actions from '../../store/actions/index';
import { Link } from 'react-router-dom';
import Spinner from '../../components/UI/Spinner/Spinner';

class StartPage extends Component {

    componentDidMount() {
        this.props.init();
    }

    render() {
        let settings = this.props.loading === true ? <Spinner /> : (
            <div className={styles.settings}>
                <Input name="amount" label="Question count" options={this.props.settings.amount} onChange={this.props.changeSetting} />
                <Input name="category" label="Category" options={this.props.settings.category.map(obj => obj.name)} onChange={this.props.changeSetting} />
                <Input name="type" label="Type" options={this.props.settings.type} onChange={this.props.changeSetting} />
                <Input name="difficulty" label="Dificulity" options={this.props.settings.difficulty} onChange={this.props.changeSetting} />
            </div>
        )

        return (
            <div className={styles.StartPage}>
                <h1 className={styles.Title}>Quizer</h1>
                <div className={styles.SubTitle}>
                    <p>Quiz generator with use of Trivia API opentdb.com</p>
                    <p>created by Jacek Smetek</p>
                </div>
                {settings}
                <Link
                    to="/quiz/1"
                    className={styles.StartButton}
                    onClick={this.props.startQuiz}>
                    Start</Link>
            </div>
        )
    }
}

const mapStateToProps = state => {
    return {
        settings: state.startPage.settings.available,
        loading: state.startPage.loading,
        apiURL: state.startPage.settings.apiURL
    }
}

const mapDispatchToProps = dispatch => {
    return {
        init: () => dispatch(actions.init()),
        startQuiz: () => dispatch(actions.generateURL()),
        changeSetting: (setting, value) => dispatch(actions.setSetting(setting, value))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(StartPage);