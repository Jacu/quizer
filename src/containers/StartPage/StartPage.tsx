import React, { Component } from 'react';
import * as styles from './StartPage.css';
import { connect } from 'react-redux';
import Input from '../../components/Input/Input';
import * as actions from '../../store/actions/index';
import Spinner from '../../components/UI/Spinner/Spinner';
import Button from '../../components/UI/Button/Button';

interface StartPageProps {
    init,
    generateURL,
    initQuiz,
    settings,
    loading,
    changeSetting,
}

class StartPage extends Component<StartPageProps, {}> {
    constructor(props){
        super(props);
        this.handleStartButtonClick = this.handleStartButtonClick.bind(this);
    }

    componentDidMount() {
        this.props.init();
    }

    handleStartButtonClick() {
        this.props.generateURL();
        this.props.initQuiz();
    }

    render() {
        console.log('styles',styles);
        const {settings, loading, changeSetting} = this.props;
        const settingInputs = loading === true ? <Spinner/> : (
            <div className={styles.settingInputs}>
                <Input name="amount" label="Question count" options={settings.amount} onChange={changeSetting} />
                <Input name="category" label="Category" options={settings.category.map(obj => obj.name)} onChange={changeSetting} />
                <Input name="type" label="Type" options={settings.type} onChange={changeSetting} />
                <Input name="difficulty" label="Dificulity" options={settings.difficulty} onChange={changeSetting} />
            </div>
        )

        return (
            <div className={styles.StartPage}>
                <div className={styles.Menu}>
                    <h1 className={styles.Title}>Quizer</h1>
                    <div className={styles.SubTitle}>
                        <p>Quiz generator with use of Trivia API opentdb.com</p>
                        <p>created by Jacek Smetek</p>
                    </div>
                    {settingInputs}
                </div>
                <Button
                    link="/quiz/1"
                    onClick={this.handleStartButtonClick}
                    label="Start" />
            </div>
        )
    }
}

const mapStateToProps = ({startPage}) => {
    return {
        settings: startPage.settings.available,
        loading: startPage.loading,
        apiURL: startPage.settings.apiURL
    }
}

const mapDispatchToProps = dispatch => {
    return {
        init: () => dispatch(actions.init()),
        generateURL: () => dispatch(actions.generateURL()),
        initQuiz: () => dispatch(actions.initQuiz()),
        changeSetting: (setting, value) => dispatch(actions.setSetting(setting, value)),
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(StartPage);