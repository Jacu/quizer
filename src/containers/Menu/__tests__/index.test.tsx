import React from 'react';
import StartPage from "../";
import { render, fireEvent } from "@testing-library/react";
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import { generateURL, initQuiz, init } from "../../../store/actions";

const initialState = {
    startPage: {
        settings: {
            apiURL: 'url',
        },
        loading: false,
        dataFetched: false,
    },
};

jest.mock('../../../store/actions', () => ({
    actionTypes: {},
    generateURL: jest.fn(() => ({ type: 'type' })),
    initQuiz: jest.fn(() => ({ type: 'type' })),
    init: jest.fn(() => ({ type: 'type' })),
}));

jest.mock('../../../components/SettingPanel', () => () => <div>MOCKED SETTING PANEL</div>);

const mockStore = configureMockStore();
const store = mockStore(initialState);

describe('StartPage', () => {
    beforeEach(() => {
        jest.clearAllMocks();
        jest.resetModules();
        store.clearActions();
    });

    it('should render with props', () => {
        const { asFragment } = render(<Provider store={store}><StartPage /></Provider>);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should init when no data', () => {
        render(<Provider store={store}><StartPage /></Provider>);
        expect(init).toHaveBeenCalledTimes(1);
    });

    it('should NOT init when data available', () => {
        const mockStoreLoading = mockStore({ startPage: { ...initialState.startPage, dataFetched: true } });
        render(<Provider store={mockStoreLoading}><StartPage /></Provider>);
        expect(init).not.toHaveBeenCalled();
    });

    it('should not render button when loading', () => {
        const mockStoreLoading = mockStore({ startPage: { ...initialState.startPage, loading: true } });
        const { queryByText } = render(<Provider store={mockStoreLoading}><StartPage /></Provider>);
        expect(queryByText('Start')).toBeNull();
    });

    it('should not render setting panel when loading', () => {
        const mockStoreLoading = mockStore({ startPage: { ...initialState.startPage, loading: true } });
        const { queryByText } = render(<Provider store={mockStoreLoading}><StartPage /></Provider>);
        expect(queryByText('MOCKED SETTING PANEL')).toBeNull();
    });

    it('should generateURL on Start click', () => {
        const { getByText } = render(<Provider store={store}><StartPage /></Provider>);
        fireEvent.click(getByText('Start'));
        expect(generateURL).toHaveBeenCalledTimes(1);
    });

    it('should initQuiz on Start click', () => {
        const { getByText } = render(<Provider store={store}><StartPage /></Provider>);
        fireEvent.click(getByText('Start'));
        expect(initQuiz).toHaveBeenCalledTimes(1);
    });

    it('should initQuiz on Enter', () => {
        const { container } = render(<Provider store={store}><StartPage /></Provider>);
        fireEvent.keyDown(container.firstChild!,{key: 'Enter', code: 'Enter'})
        expect(generateURL).toHaveBeenCalledTimes(1);
        expect(initQuiz).toHaveBeenCalledTimes(1);
    });
});