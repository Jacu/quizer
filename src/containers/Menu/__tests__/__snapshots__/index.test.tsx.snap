// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StartPage should render with props 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm dymdYj"
    tabindex="0"
  >
    <div
      class="sc-AxiKw cHoWxU"
    >
      <h1
        class="sc-AxhCb ckApzy"
      >
        Quizer
      </h1>
      <div
        class="sc-AxhUy cLpufN"
      >
        <p>
          Quiz generator with use of Trivia API opentdb.com
        </p>
        <p>
          created by Jacek Smetek
        </p>
      </div>
      <div>
        MOCKED SETTING PANEL
      </div>
    </div>
    <div
      class="sc-AxheI bYrzp"
      size="2"
    >
      Start
    </div>
  </div>
</DocumentFragment>
`;
